src/gradient_aware_harmonisation/convergence.py

```diff
@@ -0,0 +1,335 @@
+    # ]
+    # """Domain of spline"""
+
+    @overload
+    def __call__(self, x: int | float) -> int | float: ...
+
+    @overload
+    def __call__(self, x: NP_FLOAT_OR_INT) -> NP_FLOAT_OR_INT: ...
+
+    @overload
+    def __call__(self, x: NP_ARRAY_OF_FLOAT_OR_INT) -> NP_ARRAY_OF_FLOAT_OR_INT: ...
+
+    def __call__(  # noqa: PLR0911
+        self, x: int | float | NP_FLOAT_OR_INT | NP_ARRAY_OF_FLOAT_OR_INT
+    ) -> int | float | NP_FLOAT_OR_INT | NP_ARRAY_OF_FLOAT_OR_INT:
+        """
+        Evaluate the spline at a given x-value
+
+        Parameters
+        ----------
+        x
+            x-value
+
+        Returns
+        -------
+        :
+            Value of the spline at `x`
+        """
+
+        # compute weight (here: gamma) according to a cosine-decay
+        def decay(
+            x: int | float | NP_FLOAT_OR_INT | NP_ARRAY_OF_FLOAT_OR_INT,
+        ) -> int | float | NP_FLOAT_OR_INT | NP_ARRAY_OF_FLOAT_OR_INT:
+            """Get cosine-decay"""
+            angle: int | float | NP_FLOAT_OR_INT | NP_ARRAY_OF_FLOAT_OR_INT = np.divide(
+                np.pi * (x - self.initial_time), (self.final_time - self.initial_time)
+            )
+            gamma = 0.5 * (1 + np.cos(angle))
+
+            # the weighted sum for computing the harmonised AND converged
+            # function has the form: "gamma * harmonised + (1-gamma) * convergence"
+            # depending on which product we want to compute (LHS or RHS of sum),
+            # we need gamma or 1-gamma, therefore we include the following condition
+            if self.apply_to_convergence:
+                return 1 - gamma
+            return gamma
+
+        if not isinstance(x, np.ndarray):
+            if x < self.initial_time:
+                if self.apply_to_convergence:
+                    return 0.0
+                return 1.0
+
+            if x >= self.final_time:
+                if self.apply_to_convergence:
+                    return 1.0
+                return 0.0
+
+            return decay(x)
+
+        # apply decay function only to values that lie between harmonisation
+        # time and convergence-time
+        x_gte_final_time = np.where(x >= self.final_time)
+        x_decay = np.logical_and(x >= self.initial_time, x < self.final_time)
+        gamma = np.ones_like(x, dtype=float)
+        gamma[x_gte_final_time] = 0.0
+        gamma[x_decay] = decay(x[x_decay])
+
+        if self.apply_to_convergence:
+            return 1 - gamma
+
+        return gamma
+
+    def derivative(self) -> CosineDecaySplineHelperDerivative:
+        """
+        Calculate the derivative of self
+
+        Returns
+        -------
+        :
+            Derivative of self
+        """
+        return CosineDecaySplineHelperDerivative(
+            initial_time=self.initial_time,
+            final_time=self.final_time,
+            apply_to_convergence=self.apply_to_convergence,
+        )
+
+    def antiderivative(self) -> SumOfSplines:
```

Comment:
```suggestion
    def antiderivative(self) -> CosineDecaySplineHelper:
```

As above, might be clearer